### PR TITLE
Initial ocean pull

### PR DESCRIPTION
**Pull request by pkerins**
Preprocessing for 3 ocean datasets are scripted in this branch:
- ocn.005 Historical Cyclone Intensity
- ocn.007 Coral Bleaching Monitoring
- ocn.009 Sea Surface Temperature Variability

Notes:
- ocn.007 is a NRT dataset. I have scripted the process for uploading the current (daily) data as a "normal" dataset; this can presumably form the basis of a comprehensive docker script.
- One or more scripts **require** a new environment variable that specifies the directory holding the GDAL executables and Python scripts (eg `gdal_merge.py`). I tried to avoid this, but running scripts with the `subprocess` module in Windows requires specifying the location of the non-executable files (like `.py` scripts), because Windows will only use the `PATH` environment variable to automatically resolve the location of executable files (`.exe`, `.bin`, etc). I could be wrong here but this seemed the simplest way to get it to work cross-platform. So in my case, my `.env` file includes the line `GDAL_DIR="C:/Program Files/GDAL"`
- I employed two helper files to contain common code: `util_files.py` for directory and file manipulation, including GDAL stuff, and `util_cloud.py` for GCS, GEE, and AWS interactions. This means that the dataset scripts are not strictly "standalone"; however, the preprocessing scripts automatically import the modules based on their location within the repo (no pip or conda installation involved), so someone who clones the repo should be able to immediately execute the preprocessing scripts without any extra steps. This makes the scripts much shorter and will make future maintenance much simpler in the event of changes to external modules, as has already happened with GEE.
- The scripts and helper modules are (mostly) using logging rather than printing directly to console (but the log is redirected to the normal output stream).

### CODE REVIEW

**Comment by pkerins** [comment]
Oh one last thing: I know the new datasets require README files in the respective folders, but that seemed easiest to do within Github itself. I probably should have done that before the pull request. Can take care of it right after merge; let me know if you want me to proceed differently.

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    zip.extractall(data_dir)
+# in this case "processed" data file is just unzipped tif
+# but we have to rename it
+unprocessed_data_file = (os.path.join(data_dir,tifname))
+processed_data_file = os.path.join(data_dir,dataset_name+'.tif')
+print (processed_data_file)
+copyfile(unprocessed_data_file, processed_data_file)
+
+print('Uploading processed data to Google Cloud Storage.')
+gcs_uris = util_cloud.gcs_upload(processed_data_file, dataset_name, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:83]:
```suggestion
# set up Google Cloud Storage project and bucket objects
gcsClient = storage.Client(os.environ.get("CLOUDSDK_CORE_PROJECT"))
gcsBucket = gcsClient.bucket(os.environ.get("GEE_STAGING_BUCKET"))

gcs_uris = util_cloud.gcs_upload(processed_data_file, dataset_name, gcs_bucket=gcsBucket )
```

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+print('Uploading processed data to Google Earth Engine.')
+# name bands according to variable names in original netcdf
+mf_bands = [{'id': band_id, 'tileset_band_index': band_ids.index(band_id), 'tileset_id': dataset_name, 'pyramidingPolicy': pyramiding_policy} for band_id in band_ids]
+
+# Upload processed data file to GEE
+asset_name = f'projects/resource-watch-gee/{dataset_name}'
+manifest = util_cloud.gee_manifest_complete(asset_name, gcs_uris[0], mf_bands)
+task_id = util_cloud.gee_ingest(manifest, public=True)
+
+util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:94]:
```suggestion
util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcsBucket)
```

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+gcs_client = storage.Client(os.environ.get("CLOUDSDK_CORE_PROJECT"))
+gcs_bucket = gcs_client.bucket(os.environ.get("GEE_STAGING_BUCKET"))
+
+# initialize ee and eeUtil modules for uploading to Google Earth Engine
+auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
+ee.Initialize(auth)
+
+aws_bucket = 'wri-projects'
+s3_prefix = 'resourcewatch/raster/'
+
```
Comment (amsnyder) [inline ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:36]:
```suggestion
```

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+unprocessed_data_file = (os.path.join(data_dir,tifname))
+processed_data_file = os.path.join(data_dir,dataset_name+'.tif')
+print (processed_data_file)
+copyfile(unprocessed_data_file, processed_data_file)
+
+print('Uploading processed data to Google Cloud Storage.')
+gcs_uris = util_cloud.gcs_upload(processed_data_file, dataset_name, gcs_bucket=gcs_bucket)
+
+print('Uploading processed data to Google Earth Engine.')
+# name bands according to variable names in original netcdf
```
Comment (amsnyder) [inline ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:86]:
```suggestion
# initialize ee and eeUtil modules for uploading to Google Earth Engine
auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
ee.Initialize(auth)

# set pyramiding policy for GEE upload
pyramiding_policy = 'MEAN' #check
           
# name bands according to variable names in original netcdf
band_ids = ['cy_intensity', 
           ]
```

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+band_ids = ['cy_intensity', 
+           ]
+
+# set pyramiding policy for GEE upload
+pyramiding_policy = 'MEAN' #check
+
+
+# ## Execution
+
```
Comment (amsnyder) [inline ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:51]:
```suggestion
# create a new sub-directory within your specified dir called 'data'
# within this directory, create files to store raw and processed data
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    masked_tifs.append(band_masked)
+    
+logger.info('Merge masked, single-band GeoTIFFs into single, multiband GeoTIFF.')
+# merge multiple masked single-band tifs as layers in single multiband tif
+multitif = os.path.abspath(os.path.join(data_dir,dataset_name+'.tif'))
+util_files.merge_geotiffs(masked_tifs, multitif)
+processed_data_file = multitif
+
+logger.info('Uploading processed data to Google Cloud Storage.')
+gcs_uris= util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:113]:
```suggestion
# set up Google Cloud Storage project and bucket objects
gcsClient = storage.Client(os.environ.get("CLOUDSDK_CORE_PROJECT"))
gcsBucket = gcsClient.bucket(os.environ.get("GEE_STAGING_BUCKET"))

gcs_uris= util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcsBucket)
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+mf_bands = [{'id': band_id, 'tileset_band_index': band_ids.index(band_id), 'tileset_id': dataset_name, 'pyramidingPolicy': pyramiding_policy} for band_id in band_ids]
+
+# Upload processed data file to GEE
+asset_name = f'projects/resource-watch-gee/{dataset_name}'
+manifest = util_cloud.gee_manifest_complete(asset_name, gcs_uris[0], mf_bands)
+task_id = util_cloud.gee_ingest(manifest, public=True)
+
+task_id = util_cloud.gee_ingest(manifest, public=True)
+
+util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:126]:
```suggestion
util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcsBucket)
```

File: ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    val['finaltifs'] = finaltifs
+    alltifs.extend(finaltifs)
+
+logger.info('Merging masked, single-band GeoTIFFs into single, multiband GeoTIFF.')
+multitif = os.path.abspath(os.path.join(data_dir,dataset_name+'.tif'))
+util_files.merge_geotiffs(alltifs, multitif, ot='float32')
+processed_data_file = multitif
+
+logger.info('Uploading processed data to Google Cloud Storage.')
+gcs_uris = util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py:224]:
```suggestion
# set up Google Cloud Storage project and bucket objects
gcsClient = storage.Client(os.environ.get("CLOUDSDK_CORE_PROJECT"))
gcsBucket = gcsClient.bucket(os.environ.get("GEE_STAGING_BUCKET"))

gcs_uris = util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcsBucket )
```

File: ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+logger.info('Uploading processed data to Google Earth Engine.')
+# name bands according to variable names in original netcdf
+mf_bands = util_cloud.gee_manifest_bands(data_dict, dataset_name)
+# Upload processed data file to GEE
+asset_name = f'projects/resource-watch-gee/{dataset_name}'
+
+manifest = util_cloud.gee_manifest_complete(asset_name, gcs_uris[0], mf_bands)
+task_id = util_cloud.gee_ingest(manifest, public=True)
+
+util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcs_bucket)
```
Comment (amsnyder) [inline ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py:235]:
```suggestion
util_cloud.gcs_remove(gcs_uris, gcs_bucket=gcsBucket)
```

File: ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+gcs_bucket = gcs_client.bucket(os.environ.get("GEE_STAGING_BUCKET"))
+
+# initialize ee and eeUtil modules for uploading to Google Earth Engine
+auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
+ee.Initialize(auth)
+
+# amazon storage info
+aws_bucket = 'wri-projects'
+s3_prefix = 'resourcewatch/raster/'
+
```
Comment (amsnyder) [inline ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py:167]:
```suggestion
# create a new sub-directory within your specified dir called 'data'
# within this directory, create files to store raw and processed data
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+# initialize ee and eeUtil modules for uploading to Google Earth Engine
+auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
+ee.Initialize(auth)
+
+# amazon storage info
+aws_bucket = 'wri-projects'
+s3_prefix = 'resourcewatch/raster/'
+
+# ## Execution
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:76]:
```suggestion
# create a new sub-directory within your specified dir called 'data'
# within this directory, create files to store raw and processed data
```

File: ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+logger.info('Merging masked, single-band GeoTIFFs into single, multiband GeoTIFF.')
+multitif = os.path.abspath(os.path.join(data_dir,dataset_name+'.tif'))
+util_files.merge_geotiffs(alltifs, multitif, ot='float32')
+processed_data_file = multitif
+
+logger.info('Uploading processed data to Google Cloud Storage.')
+gcs_uris = util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcs_bucket)
+
+logger.info('Uploading processed data to Google Earth Engine.')
+# name bands according to variable names in original netcdf
```
Comment (amsnyder) [inline ocn_007_coral_bleaching_monitoring/ocn_007_coral_bleaching_monitoring.py:227]:
```suggestion
# initialize ee and eeUtil modules for uploading to Google Earth Engine
auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
ee.Initialize(auth)

# name bands according to variable names in original netcdf
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+# merge multiple masked single-band tifs as layers in single multiband tif
+multitif = os.path.abspath(os.path.join(data_dir,dataset_name+'.tif'))
+util_files.merge_geotiffs(masked_tifs, multitif)
+processed_data_file = multitif
+
+logger.info('Uploading processed data to Google Cloud Storage.')
+gcs_uris= util_cloud.gcs_upload(multitif, dataset_name, gcs_bucket=gcs_bucket)
+
+logger.info('Uploading processed data to Google Earth Engine.')
+# name bands according to variable names in original netcdf
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:116]:
```suggestion
# initialize ee and eeUtil modules for uploading to Google Earth Engine
auth = ee.ServiceAccountCredentials(os.getenv('GEE_SERVICE_ACCOUNT'), os.getenv('GOOGLE_APPLICATION_CREDENTIALS'))
ee.Initialize(auth)

# set pyramiding policy for GEE upload
pyramiding_policy = 'MEAN' #check
           
# name bands according to variable names in original netcdf
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+               'mask',
+              ]
+
+band_ids = ['stdv_maxmonth',
+            'stdv_annual',
+           ]
+
+# set pyramiding policy for GEE upload
+pyramiding_policy = 'MEAN' #check
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:49]:
```suggestion
```

File: ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py (hunk 2 above)
Comment (pkerins) [reply ocn_005_historical_cyclone_intensity/ocn_005_historical_cyclone_intensity.py:94]: Accepting change. [Strictly FYI](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names).

**Comment by pkerins** [comment]
I think all comments/requests have been addressed/accepted. Will pull and test scripts.

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+logger.addHandler(console)
+logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+
+logger.info('Executing script for dataset: ' + dataset_name)
+# create a new sub-directory within your specified dir called 'data'
+# within this directory, create files to store raw and processed data
+data_dir = util_files.prep_dirs(dataset_name)
+logger.debug('Data directory relative path: '+data_dir)
+logger.debug('Data directory absolute path: '+os.path.abspath(data_dir))
+logger.info('Downloading raw data')
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:64]:
```suggestion
'''
Download data and save to your data directory
'''
logger.info('Downloading raw data')
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+url='ftp://ftp.star.nesdis.noaa.gov/pub/sod/mecb/crw/data/thermal_history/v2.1/noaa_crw_thermal_history_sst_variability_v2.1.nc'
+raw_data_file = os.path.join(data_dir,os.path.basename(url))
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:67]:
```suggestion
# insert the url used to download the data from the source website
url='ftp://ftp.star.nesdis.noaa.gov/pub/sod/mecb/crw/data/thermal_history/v2.1/noaa_crw_thermal_history_sst_variability_v2.1.nc'

# download the data from the source
raw_data_file = os.path.join(data_dir,os.path.basename(url))
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+urllib.request.urlretrieve(url, raw_data_file)
+
+logger.debug('Raw data file path: ' + raw_data_file)
+
+logger.info('Extracting relevant GeoTIFFs from source NetCDF')
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:72]:
```suggestion
'''
Process data
'''
logger.info('Extracting relevant GeoTIFFs from source NetCDF')
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:73]:
```suggestion
# convert netcdf to individual tif files for each of the subdatasets specified
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+               'stdv_annual',
+               'mask',
+              ]
+
+# nedcdf subdatasets that will be uploaded as bands to GEE
+band_ids = ['stdv_maxmonth',
+            'stdv_annual',
+           ]
+# ### Set/initialize/authenticate general variables
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:48]:
```suggestion
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+# Get the top-level logger object
+logger = logging.getLogger()
+for handler in logger.handlers: logger.removeHandler(handler)
+logger.setLevel(logging.INFO)
+# make it print to the console.
+console = logging.StreamHandler()
+logger.addHandler(console)
+logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:57]:
```suggestion
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+import urllib
+from zipfile import ZipFile
+import ee
+import subprocess
+from google.cloud import storage
+from shutil import copyfile
+import time
+import logging
+
+
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:30]:
```suggestion

# Get the top-level logger object
logger = logging.getLogger()
for handler in logger.handlers: logger.removeHandler(handler)
logger.setLevel(logging.INFO)
# make it print to the console.
console = logging.StreamHandler()
logger.addHandler(console)
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
```

File: ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+tifs = util_files.convert_netcdf(raw_data_file, subdatasets)
+
+logger.info('Mask extracted GeoTIFFs using dataset mask')
+# create dict linking subdatasets from the netcdf to the geotiffs that now contain each
+sds_file_dict=dict(zip(subdatasets,tifs))
+# set nodata value for masking
+nodata=-128
+
+# cycle through target tifs (corresponding to bands) and mask them with mask tif
+masked_tifs = []
```
Comment (amsnyder) [inline ocn_009_sea_surface_temperature_variability/ocn_009_sea_surface_temperature_variability.py:83]:
```suggestion
# netcdf subdatasets that will be uploaded as bands to GEE
band_ids = ['stdv_maxmonth',
            'stdv_annual',
           ]
           
masked_tifs = []
```

**Comment by pkerins** [comment]
Scripts for `ocn.005`, `ocn.007`, and `ocn.009` all confirmed to work as committed.

**Pull request merged**